core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -352,7 +352,14 @@ class LogSegment(val log: FileMessageSet,
         // If the log is empty, we return time elapsed since the segment is created.
         return now - created
```

Comment:
I think the logic would be clearer if we removed this code here and changed the code at the end of the method to be like:

```scala
rollingBasedTimestamp match {
  case Some(t) if t >= 0 => messageTimestamp - t
  case _ => now - created
```